staging/src/k8s.io/apiserver/pkg/server/width/width.go

```diff
@@ -0,0 +1,91 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package width
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// DefaultRequestWidthFinder returns '1' as the width of any
+// given request, this is in keeping with the current behavior.
+var DefaultRequestWidthFinder = &defaultWidthFinder{}
+
+// NewChainedRequestWidthFinder returns a RequestWidthFinder that executes
+// the given chain of RequestWidthFinder(s) one at a time in the specified order
+// and returns the "width" of a given request.
+// It stops when a RequestWidthFinder from the chain has returned the "width"
+// of the given request.
+// If no RequestWidthFinder from the chain returned a "width" for the given request
+// the the chain is exhausted and we fallback to the default finder which returns
+// '1' as the "width" of the given request.
+func NewChainedRequestWidthFinder(chain ...RequestWidthFinder) RequestWidthFinder {
+	return &chainedWidthFinder{
+		fallback: &defaultWidthFinder{},
+		chain:    chain,
+	}
+}
+
+// RequestWidthFinder finds the "width" of the given request.
+type RequestWidthFinder interface {
+	// Name returns the name of the RequestWidthFinder, it can be useful in
+	// logging error, for example if we have a chain of RequestWidthFinder(s)
+	// we can include the name of the RequestWidthFinder in the error log.
+	Name() string
+
+	// FindWidth calculates the "width" of a given request.
+	//
+	// err is set when the function ran into an error, this indicates
+	// that we should abort the request.
+	// ok is 'true' when the function successfully calculated the
+	// "width" of the request, otherwise it is set to 'false'
+	// width would be set to a non-zero value when the function
+	// successfully calculated the "width", otherwsie it is set to zero.
+	FindWidth(r *http.Request) (width float64, ok bool, err error)
+}
+
+type defaultWidthFinder struct{}
+
+func (defaultWidthFinder) Name() string {
+	return "default"
+}
+func (defaultWidthFinder) FindWidth(r *http.Request) (width float64, ok bool, err error) {
+	return 1, true, nil
+}
+
+type chainedWidthFinder struct {
+	chain    []RequestWidthFinder
+	fallback RequestWidthFinder
+}
+
+func (c *chainedWidthFinder) Name() string {
+	return "chained"
+}
+
+func (c *chainedWidthFinder) FindWidth(r *http.Request) (float64, bool, error) {
+	for _, finder := range c.chain {
+		width, ok, err := finder.FindWidth(r)
+		if err != nil {
+			return 0, false, fmt.Errorf("failed to calculate width of request, error from %s - %w", finder.Name(), err)
+		}
+		if ok {
+			return width, true, nil
```

Comment:
OK - I see why you want to have "ok". 

So IIUC, you eventually want to have N different finders, e.g. list one, watch one, ... and combine the value from there.
I think this is overenginnering and until we really need something like that, I would avoid that. [I think that for now we can simply have separate functions for all of those and FindWidth just dispatching to the right now.

Reply:
yes, we will need some sort of dispatching mechanism if we have more than 1, I removed it for now.

Reply:
when we plumb the actual "width" we will need to handle error like this below:
```
func(r *http.Request) float64 {
		for _, estimator := range chain {
			width, ok, err := estimator.EstimateWidth(r)
			if err != nil {
				klog.ErrorS(err, "Ran into an error while estimating width", "name", estimator.Name())
				break
			}
			if ok {
				return width
			}
		}

		// return the default width
		return 1.0
	}
```